File: application/src/components/Forms/src/com/eas/client/forms/api/Orientation.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package com.eas.client.forms.api;

import com.eas.script.AlreadyPublishedException;
import com.eas.script.HasPublished;
import com.eas.script.NoPublisherException;
import com.eas.script.ScriptFunction;
import com.eas.script.ScriptObj;
import jdk.nashorn.api.scripting.JSObject;

/**
 *
 * @author mg
 */
@ScriptObj(jsDoc = "/**\n"
        + "* Orientation types constants.\n"
        + "*/")
public class Orientation implements HasPublished {

    public static final int HORIZONTAL = 0;
    public static final int VERTICAL = 1;
    
    protected static JSObject publisher;
    protected Object published;
    
    @ScriptFunction
    public static int getHORIZONTAL() {
        return HORIZONTAL;
    }
    
    @ScriptFunction
    public static int getVERTICAL() {
        return VERTICAL;
    }
    
    @Override
    public Object getPublished() {
        if (published == null) {
            if (publisher == null || !publisher.isFunction()) {
                throw new NoPublisherException();
            }
            published = publisher.call(null, new Object[]{});
        }
        return published;
    }

    @Override
    public void setPublished(Object aValue) {
        if (published != null) {
            throw new AlreadyPublishedException();
        }
        published = aValue;
    }

    public static void setPublisher(JSObject aPublisher) {
        publisher = aPublisher;
    }
}

File: application/api/forms/orientation.js
(function() {
    var javaClass = Java.type("com.eas.client.forms.api.Orientation");
    javaClass.setPublisher(function(aDelegate) {
        return new P.Orientation(aDelegate);
    });
    
    /**
     * Generated constructor.
     * @constructor Orientation Orientation
     */
    P.Orientation = function () {

        var maxArgs = 0;
        var delegate = arguments.length > maxArgs ?
            arguments[maxArgs] : new javaClass();

        Object.defineProperty(this, "unwrap", {
            get: function() {
                return function() {
                    return delegate;
                };
            }
        });
        /**
         * Generated property jsDoc.
         * @property VERTICAL
         * @memberOf Orientation
         */
        Object.defineProperty(this, "VERTICAL", {
            get: function() {
                var value = delegate.VERTICAL;
                return P.boxAsJs(value);
            }
        });

        /**
         * Generated property jsDoc.
         * @property HORIZONTAL
         * @memberOf Orientation
         */
        Object.defineProperty(this, "HORIZONTAL", {
            get: function() {
                var value = delegate.HORIZONTAL;
                return P.boxAsJs(value);
            }
        });


        delegate.setPublished(this);
    };
})();